tests/test_assets.py

```diff
@@ -940,63 +951,126 @@ def test_map_identifier_index_to_sids(self):
+        sids = np.arange(len(equities))
+        np.random.RandomState(1337).shuffle(sids)
+        equities.index = sids
+        permute_sid = dict(zip(sids, range(len(sids)))).__getitem__
+
+        exchanges = pd.DataFrame.from_records([
+            {'exchange': 'US_EXCHANGE_1', 'country_code': 'US'},
+            {'exchange': 'US_EXCHANGE_2', 'country_code': 'US'},
+            {'exchange': 'CA_EXCHANGE', 'country_code': 'CA'},
+            {'exchange': 'JP_EXCHANGE', 'country_code': 'JP'},
+        ])
+        sids_per_country = {
+            'US': equities.index[:2 * assets_per_exchange],
+            'CA': equities.index[
+                2 * assets_per_exchange:3 * assets_per_exchange
+            ],
+            'JP': equities.index[3 * assets_per_exchange:],
+        }
+        self.write_assets(equities=equities, exchanges=exchanges)
         finder = self.asset_finder
 
         all_dates = pd.date_range(
             start=first_start,
-            end=frame.end_date.max(),
+            end=equities.end_date.max(),
             freq=trading_day,
         )
 
         for dates in all_subindices(all_dates):
             expected_with_start_raw = full(
-                shape=(len(dates), num_assets),
+                shape=(len(dates), assets_per_exchange),
                 fill_value=False,
                 dtype=bool,
             )
             expected_no_start_raw = full(
-                shape=(len(dates), num_assets),
+                shape=(len(dates), assets_per_exchange),
                 fill_value=False,
                 dtype=bool,
             )
 
             for i, date in enumerate(dates):
-                it = frame[['start_date', 'end_date']].itertuples()
-                for j, start, end in it:
+                it = equities.iloc[:4][['start_date', 'end_date']].itertuples(
+                    index=False,
+                )
+                for j, (start, end) in enumerate(it):
                     # This way of doing the checks is redundant, but very
                     # clear.
                     if start <= date <= end:
                         expected_with_start_raw[i, j] = True
                         if start < date:
                             expected_no_start_raw[i, j] = True
 
-            expected_with_start = pd.DataFrame(
-                data=expected_with_start_raw,
-                index=dates,
-                columns=frame.index.values,
-            )
-            result = finder.lifetimes(dates, include_start_date=True)
-            assert_frame_equal(result, expected_with_start)
-
-            expected_no_start = pd.DataFrame(
-                data=expected_no_start_raw,
-                index=dates,
-                columns=frame.index.values,
-            )
-            result = finder.lifetimes(dates, include_start_date=False)
-            assert_frame_equal(result, expected_no_start)
+            for country_codes in powerset(exchanges.country_code.unique()):
```

Comment:
I don't really want to duplicate this too much because this test is _really_ slow. I think it is probably better to more thoroughly explain this with comments.